01-resources-v2/proposal.md

```diff
@@ -0,0 +1,285 @@
+to them on stdin. I'm going to use the `Bits` type just to represent which calls
+have state on disk to either access (like `put`) or return (like `get`). This
+would normally just be the working directory of the script.
+
+## General Types
+
+```elm
+-- arbitrary configuration
+type alias Config = Dict String Json.Value
+
+-- identifier for space, i.e. {"branch":"foo"}
+type alias Space = Dict String String
+
+-- identifier for version, i.e. {"version":"1.2"}
+type alias Version = Dict String String
+
+-- arbitrary ordered metadata (we may make this fancier in the future)
+type alias Metadata = List (String, String)
+
+-- data on disk
+type alias Bits = ()
+```
+
+## Versioned Artifacts interface
+
+```elm
+spaces  : Config -> Set Space
+check   : Config -> Space -> Maybe Version -> List (Version, Metadata)
+get     : Config -> Space -> Version -> Bits
+put     : Config -> Bits -> (Space, Set Version)
+destroy : Config -> Bits -> Dict Space (Set Version)
+```
+
+## Notifications interface
+
+```elm
+notify : Config -> Notification -> ()
+
+type Notification
+  = BuildStarted
+      { build : BuildMetadata
+      , self : Maybe BuildInput
+      }
+  | BuildFinished
+      { build : BuildMetadata
+      , status : Status
+      , self : Maybe BuildInput
+      }
+
+type alias BuildMetadata =
+  { teamName : String
+  , pipelineName : String
+  , jobName : String
+  , buildName : String
+  , buildID : Int
+  , status : String
+  }
+
+type alias BuildInput =
+  { space : Space
+  , version : Version
+  }
+```
+
+
+# Examples
+
+TODO:
+
+- Pull Requests
+- Feature branches
+- Build matrixes
+- Generating branches (and propagating them downstream)
+- Semver artifacts
+- Fanning out against multiple IaaSes
+- Pool resource?
+- BOSH deploys
+
+
+# Summary of Changes
+
+## Overarching Changes
+
+* Add a `info` script which prints the resource's API version, e.g.
+  `{"version":"2.0"}`. This will start at `2.0`. If `/info` does not exist we'll
+  execute today's resource interface behavior.
+
+* Rather than running `/opt/resource/X`, discover the supported resource APIs by
+  invoking `info`. This allows us to be more flexible in what kinds of resources
+  we can support (versioned artifacts, notifications, ???), and where the
+  scripts may live (`/opt/resource` is very Linux specific).
+
+* Today's resource interface (`/in`, `/out`, `/check`) becomes more specifically
+  a "versioned artifacts" or just "artifacts" resource interface.
+
+* Introduction of some sort of schema validation for resource configuration.
+
+* Remove the distinction between `source` and `params`; resources will receive a
+  single `config`. The distinction will remain in the pipeline. This makes it
+  easier to implement a resource without planning ahead for interesting dynamic
+  vs. static usage patterns, and will get more powerful with #684.
+
+
+## Changes to Versioned Artifact resources
+
+* Add a `spaces` actopm, which is used to discover spaces.
```

Comment:
action? your right hand shifted left one key

Reply:
Haha, I'll fix it.